providers/oidc.go

```diff
@@ -233,3 +235,33 @@ func (p *OIDCProvider) createSession(ctx context.Context, token *oauth2.Token, r
 
 	return ss, nil
 }
+
+func (p *OIDCProvider) GetLoginURL(redirectURI, state string) string {
+	s := makeLoginURL(p.ProviderData, redirectURI, state, p.makeIDTokenClaims())
+	return s.String()
+}
+
+func (p *OIDCProvider) makeIDTokenClaims() url.Values {
```

Comment:
Interesting, learned something new today 😄 

Looking at the documentation you linked, it seems like there's much more variety in what this parameter can have and we've honed into a very specific subset:

```
  {
   "userinfo":
    {
     "given_name": {"essential": true},
     "nickname": null,
     "email": {"essential": true},
     "email_verified": {"essential": true},
     "picture": null,
     "http://example.info/claims/groups": null
    },
   "id_token":
    {
     "auth_time": {"essential": true},
     "acr": {"values": ["urn:mace:incommon:iap:silver"] }
    }
  }
```

I think it may be easier for us on the long-term maintenance front and power users that really want to customize this if we just take in the raw JSON we want in the `claims` parameter to the `authentication_endpoint`.

We can just document sample use cases in the documentation to help new users get started with the feature?

Reply:
I totally understand the hesitation. Your configuration does not quite match with these complex structures. That was the reason, we picked only a small part, we needed to get at least `groups` and `affiliation` in the ID Token from Satosa.

The downside of only documenting the whole think, is that you have to manually set all urls in your configuration and I am also not quite sure, if this is easier maintainable in the future, since those kind of things often become broken during updates, since nobody thinks about those special cases, where someone has added url parameter to the ´login-url`.

I would have 3 different suggestions:

1. Use this code as it is. I think this would be the default use case. To add new claims to the ID Token.
2. Add at least also the `authinfo`, so that you can give a list of `authinfo` and `id_token` claims and support something like
```json
{
  "userinfo":
  {
    "given_name": {"essential": true},
    "email": {"essential": true},
    "email_verified": {"essential": true}
   },
   "id_token":
   {
     "auth_time": {"essential": true}
   }
}
```
3. Only document the use case.

Of course we could still document the remaining cases in suggestion 1 and 2.

Reply:
I'm strongly leaning toward total flexibility (and simpler code 😉 ) for these two reason:

(1) We are moving to structured configs in v8 (you can see the beginnings of this in the alpha config). It will be way easier to put in complex structures as a config value then (where it is a little tedious now with config flags).
(2) My goal is to have the OIDCProvider be the base of alot of other providers via embedding. Leaving the implementation open ended on the OIDCProvider will allow any more specific implementations of OIDC to have the power to set it exactly as they need.

Reply:
Okay, so do I understand this correct, that this code will not be accepted?

What about adding this feature to the alpha version? Would that be an option?

Reply:
@JoelSpeed What are your thoughts?

I want to keep support for the `claims` parameter in the Login URL we construct so we can add support for more aspects of the OIDC spec -- that feels like a nice addition.

But I don't want to be opinionated/prescriptive since our OIDCProvider is consumed by all sorts of very different providers and needs to be as generic as possible (vs every other provider in our codebase is tied to a specific 3rd party provider entity).

In the past, whenever we've accepted opinionated design PRs on the OIDC design -- we inevitably got more follow on PRs asking to generalize because the selected architecture did not fit with what another person's OIDC IdP supported (I feel like we had at least 4 `groups` related PRs because IdPs are all over the place).

Hence, I strongly want to give flexibility to operators if they know they need an OIDC claims JSON, to have all the power at their fingertips for the nuances of their particular IdP.

Reply:
Soo?

Reply:
I'm open to this change if we go the more extensible route. I don't want to be opinionated in the `OIDCProvider` since it is connected to such a wide variety of IdPs (vs the other OAuth based providers can be tuned to the exact provider).

The full structure might fit better in `v8` where we have structured configs if you have apprehensions about full raw JSON in config flags or envvars at the moment if you want to wait or close.

Reply:
So just catching up on this, I didn't know this even existed! 

A few questions:
- What happens if a provider doesn't support this?
- We plan to eventually allow users to request arbitrary claims to be transferred from ID Token/userinfo to our session object, could this fit in with that maybe?
- This looks to be backwards compatible at the moment, was that intentional?
- What happens if the user sets this list and we mark specific claims essential, but then the IDP doesn't support that particular claim?
- Is there a way we can request claims, but make them optional? (Not including the essential I guess?)
- Have you got any examples of providers who document that they support this feature of OIDC?

Reply:
I am no expert, but I will try to answer your questions

>     * What happens if a provider doesn't support this?

I think this depends on the provider. I personally think in most cases a URL parameter which is set, but not supported is ignored.

>     * We plan to eventually allow users to request arbitrary claims to be transferred from ID Token/userinfo to our session object, could this fit in with that maybe?

If you are not using the Redis database, these requested claims would directly be added to the session cookie.

>     * This looks to be backwards compatible at the moment, was that intentional?

I intend to add this in a way to make it possible to work perfectly fine without the new config parameter set.

>     * What happens if the user sets this list and we mark specific claims essential, but then the IDP doesn't support that particular claim?

From the [spec](https://openid.net/specs/openid-connect-core-1_0.html#IndividualClaimsRequests): *Note that even if the Claims are not available because the End-User did not authorize their release or they are not present, the Authorization Server MUST NOT generate an error when Claims are not returned, whether they are Essential or Voluntary, unless otherwise specified in the description of the specific claim. *

>     * Is there a way we can request claims, but make them optional? (Not including the essential I guess?)

Yes, take again a look at the  [spec](https://openid.net/specs/openid-connect-core-1_0.html#IndividualClaimsRequests) for this

> 
>     * Have you got any examples of providers who document that they support this feature of OIDC?

[Satosa](https://github.com/IdentityPython/SATOSA) supports this.